akka-cluster/src/test/scala/akka/cluster/JoinConfigCompatCheckerRollingUpdateSpec.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster
+
+import scala.concurrent.duration._
+import scala.collection.{ immutable ⇒ im }
+import scala.util.Random
+
+import akka.actor.ExtendedActorSystem
+import akka.testkit.{ AkkaSpec, LongRunningTest }
+import com.typesafe.config.{ Config, ConfigFactory }
+import org.scalatest.DoNotDiscover
+
+object JoinConfigCompatCheckerRollingUpdateSpec {
+
+  val baseConfig = ConfigFactory.parseString(
+    s"""
+      akka.loglevel = WARNING
+      akka.log-dead-letters = off
+      akka.log-dead-letters-during-shutdown = off
+      akka.remote.log-remote-lifecycle-events = off
+      akka.cluster.jmx.enabled = off
+    """)
+    .withFallback(JoinConfigCompatCheckerSpec.baseConfig)
+
+  val v1Config: Config = baseConfig.withFallback(JoinConfigCompatCheckerSpec.configWithChecker)
+
+  val v2Config: Config = ConfigFactory.parseString(
+    """
+      akka.cluster.v2-configuration = "v2"
+      akka.cluster.configuration-compatibility-check.checkers {
+        rolling-upgrade-test = "akka.cluster.JoinConfigCompatRollingUpdateChecker"
+      }
+    """).withFallback(v1Config)
+
+  def unenforced(config: Config): Config =
+    ConfigFactory
+      .parseString("""akka.cluster.configuration-compatibility-check.enforce-on-join = off""")
+      .withFallback(config)
+}
+
+class JoinConfigCompatCheckerDiffSpec extends AkkaSpec {
+
+  "JoinConfigCompatChecker" must {
+
+    val sys = system.asInstanceOf[ExtendedActorSystem]
+    val defaults = ConfigFactory.load().withOnlyPath("akka.cluster")
+      .withFallback(ConfigFactory.parseString("""akka.version = "2.5.9"""")) // should have from load
+
+    val v1Config = JoinConfigCompatCheckerRollingUpdateSpec.v1Config.withFallback(defaults)
+    val v2Config = JoinConfigCompatCheckerRollingUpdateSpec.v2Config.withFallback(defaults)
+
+    "diff ValidNel if the joining config has new, additional properties and has all required keys" in {
+      val composite = JoinConfigCompatChecker.load(sys, new ClusterSettings(v1Config, system.name))
+      val result = JoinConfigCompatChecker.diff(composite.requiredKeys, v2Config, v1Config)
+      result should ===(ValidNel)
+    }
+
+    "diff Valid if the joining config has no new properties and has all required keys" in {
+      val composite = JoinConfigCompatChecker.load(sys, new ClusterSettings(v1Config, system.name))
+      val result = JoinConfigCompatChecker.diff(composite.requiredKeys, v1Config, v1Config)
+      result should ===(Valid)
+    }
+
+    "be Valid in full check during a rolling upgrade if the joining config has new, additional properties and is otherwise compatible" in {
+      val composite = JoinConfigCompatChecker.load(sys, new ClusterSettings(v1Config, system.name))
+      val result = JoinConfigCompatChecker.fullMatch(composite.requiredKeys, v2Config, v1Config)
```

Comment:
I'm not sure these tests do the same thing as a real cluster.
See `joinConfigCompatChecker.check(joiningNodeConfig, configWithoutSensitiveKeys)` and `joinConfigCompatChecker.check(clusterConfig, context.system.settings.config)`